File: .NET/Examples/WindowsDesktop/Components/PrimaryInterfacePackageControl.cs
using System;
using System.ComponentModel;
using System.Globalization;
using System.Linq;
using System.Windows.Forms;
using UnderAutomation.UniversalRobots;
using UnderAutomation.UniversalRobots.Common;
using UnderAutomation.UniversalRobots.PrimaryInterface;

// Advanced Property Grid thats displays a Primary Interface Package
public partial class PrimaryInterfacePackageControl : UserControl
{
    private VScrollBar _vscroll;

    static PrimaryInterfacePackageControl()
    {
        // make grid objects expandable
        TypeDescriptor.AddAttributes(typeof(JointData), new TypeConverterAttribute(typeof(CustomExpandableObjectConverter)));
        TypeDescriptor.AddAttributes(typeof(JointConfiguration), new TypeConverterAttribute(typeof(CustomExpandableObjectConverter)));
        TypeDescriptor.AddAttributes(typeof(JointKinematicsInfo), new TypeConverterAttribute(typeof(CustomExpandableObjectConverter)));
        TypeDescriptor.AddAttributes(typeof(MasterboardDigitalIO), new TypeConverterAttribute(typeof(CustomExpandableObjectConverter)));

        // make grid readonly
        TypeDescriptor.AddAttributes(typeof(PackageEventArgs), new ReadOnlyAttribute(true));
    }

    public PrimaryInterfacePackageControl()
    {
        InitializeComponent();
        _vscroll = grid.Controls.OfType<Control>().FirstOrDefault(ctl => ctl.AccessibilityObject.Role == AccessibleRole.Table)?.Controls?.OfType<VScrollBar>()?.FirstOrDefault();
    }


    public string LabelText
    {
        get
        {
            return lbl.Text;
        }
        set
        {
            lbl.Text = value;
        }
    }

    public void SetSelectedObject<T>(T value) where T: PackageEventArgs
    {
        this.SuspendLayout();

        // conserve scroll bar position after a refresh
        var vScrollValue = _vscroll?.Value;

        if (value is null)
        {
            if (grid.SelectedObject is null) grid.SelectedObject = Activator.CreateInstance<T>();
        }
        else
        {
            grid.SelectedObject = value;
            grid.ExpandAllGridItems();

            if (_vscroll != null)
                _vscroll.Value = vScrollValue ?? 0;
        }
        this.ResumeLayout();
    }

    public class CustomExpandableObjectConverter : ExpandableObjectConverter
    {
        public override object ConvertTo(ITypeDescriptorContext context, CultureInfo culture, object value, Type destinationType)
        {
            // Return empty string and not class name in data grid view for expandable objects
            return destinationType == typeof(string) ? "" : base.ConvertTo(context, culture, value, destinationType);
        }
    }
}

File: UnderAutomation.UniversalRobots.Showcase.Forms.vb/Components/PrimaryInterfacePackageControl.vb
Imports System
Imports System.ComponentModel
Imports System.Globalization
Imports System.Linq
Imports System.Windows.Forms
Imports UnderAutomation.UniversalRobots.Common
Imports UnderAutomation.UniversalRobots.PrimaryInterface

' Advanced Property Grid thats displays a Primary Interface Package
Public Partial Class PrimaryInterfacePackageControl
    Inherits UserControl
    Private _vscroll As VScrollBar

    Shared Sub New()
        ' make grid objects expandable
        Call TypeDescriptor.AddAttributes(GetType(JointData), New TypeConverterAttribute(GetType(CustomExpandableObjectConverter)))
        Call TypeDescriptor.AddAttributes(GetType(JointConfiguration), New TypeConverterAttribute(GetType(CustomExpandableObjectConverter)))
        Call TypeDescriptor.AddAttributes(GetType(JointKinematicsInfo), New TypeConverterAttribute(GetType(CustomExpandableObjectConverter)))
        Call TypeDescriptor.AddAttributes(GetType(MasterboardDigitalIO), New TypeConverterAttribute(GetType(CustomExpandableObjectConverter)))

        ' make grid readonly
        Call TypeDescriptor.AddAttributes(GetType(PackageEventArgs), New ReadOnlyAttribute(True))
    End Sub

    Public Sub New()
        InitializeComponent()
        _vscroll = grid.Controls.OfType(Of Control)().FirstOrDefault(Function(ctl) ctl.AccessibilityObject.Role = AccessibleRole.Table)?.Controls?.OfType(Of VScrollBar)()?.FirstOrDefault()
    End Sub


    Public Property LabelText As String
        Get
            Return lbl.Text
        End Get
        Set(value As String)
            lbl.Text = value
        End Set
    End Property

    Public Sub SetSelectedObject(Of T As PackageEventArgs)(value As T)
        SuspendLayout()

        ' conserve scroll bar position after a refresh
        Dim vScrollValue = _vscroll?.Value

        If value Is Nothing Then
            If grid.SelectedObject Is Nothing Then grid.SelectedObject = Activator.CreateInstance(Of T)()
        Else
            grid.SelectedObject = value
            grid.ExpandAllGridItems()

            If _vscroll IsNot Nothing Then _vscroll.Value = If(vScrollValue, 0)
        End If
        ResumeLayout()
    End Sub

    Public Class CustomExpandableObjectConverter
        Inherits ExpandableObjectConverter
        Public Overrides Function ConvertTo(context As ITypeDescriptorContext, culture As CultureInfo, value As Object, destinationType As Type) As Object
            ' Return empty string and not class name in data grid view for expandable objects
            Return If(destinationType Is GetType(String), "", MyBase.ConvertTo(context, culture, value, destinationType))
        End Function
    End Class
End Class
